vm/systemSmartContracts/auction.go

```diff
@@ -0,0 +1,748 @@
+	}
+
+	if stakedData.Staked || stakedData.UnStakedNonce < stakedData.StartNonce {
+		log.Debug("unBound is not possible for address which is staked or is not in unbound period")
+		return nil, vm.ErrStillInUnBoundPeriod
+	}
+
+	currentNonce := s.eei.BlockChainHook().CurrentNonce()
+	if currentNonce-stakedData.UnStakedNonce < s.unBoundPeriod {
+		log.Debug("unBound is not possible for address because unbound period did not pass")
+		return nil, vm.ErrStillInUnBoundPeriod
+	}
+
+	s.eei.SetStorage(blsKey, nil)
+	config := s.getConfig(stakedData.UnStakedEpoch)
+
+	return config.NodePrice, nil
+}
+
+func (s *stakingAuctionSC) unBound(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	registrationData, err := s.getRegistrationData(args.CallerAddr)
+	if err != nil {
+		return vmcommon.UserError
+	}
+
+	blsKeys, err := getBLSPublicKeys(registrationData, args)
+	if err != nil {
+		return vmcommon.UserError
+	}
+
+	totalUnBound := big.NewInt(0)
+	for _, blsKey := range blsKeys {
+		nodePrice, err := s.deleteBLSStakedData(blsKey)
+		if err != nil {
+			return vmcommon.UserError
+		}
+
+		ownerAddress := s.eei.GetStorage([]byte(ownerKey))
+		err = s.eei.Transfer(args.CallerAddr, ownerAddress, nodePrice, nil)
+		if err != nil {
+			log.Debug("transfer error on unBound function",
+				"error", err.Error(),
+			)
+			return vmcommon.UserError
+		}
+
+		_ = totalUnBound.Add(totalUnBound, nodePrice)
+	}
+
+	if registrationData.BlockedStake.Cmp(totalUnBound) < 0 {
+		log.Debug("too much to unbound, not enough total stake")
+		return vmcommon.UserError
+	}
+
+	_ = registrationData.BlockedStake.Sub(registrationData.BlockedStake, totalUnBound)
+	_ = registrationData.TotalStakeValue.Sub(registrationData.TotalStakeValue, totalUnBound)
+
+	zero := big.NewInt(0)
+	if registrationData.BlockedStake.Cmp(zero) == 0 && registrationData.TotalStakeValue.Cmp(zero) == 0 {
+		s.eei.SetStorage(args.CallerAddr, nil)
+	} else {
+		err := s.saveRegistrationData(args.CallerAddr, registrationData)
+		if err != nil {
+			log.Debug("cannot save registration data change")
+			return vmcommon.UserError
+		}
+	}
+
+	return vmcommon.Ok
+}
+
+func (s *stakingAuctionSC) claim(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	registrationData, err := s.getRegistrationData(args.CallerAddr)
+	if err != nil {
+		return vmcommon.UserError
+	}
+
+	if len(registrationData.RewardAddress) == 0 {
+		return vmcommon.UserError
+	}
+
+	zero := big.NewInt(0)
+	claimable := big.NewInt(0).Sub(registrationData.TotalStakeValue, registrationData.BlockedStake)
+	if claimable.Cmp(zero) <= 0 {
+		return vmcommon.UserError
+	}
+
+	registrationData.TotalStakeValue.Set(registrationData.BlockedStake)
+	err = s.saveRegistrationData(args.CallerAddr, registrationData)
+	if err != nil {
+		log.Debug("cannot save registration data change")
+		return vmcommon.UserError
+	}
+
+	ownerAddress := s.eei.GetStorage([]byte(ownerKey))
+	err = s.eei.Transfer(args.CallerAddr, ownerAddress, claimable, nil)
```

Comment:
This order should be changed. Otherwise the caller ends up sending wallet funds toward the owner of this SC.

Reply:
actually it was correct as the first is the destination. :P